MusicNotationKit/MusicNotationKit/Tuplet.swift

```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
         return try note(from: fullIndexes)
     }
 
-    public mutating func replaceNote(at index: Int, with note: Note) throws {
-        try replaceNote(at: index, with: note as NoteCollection)
+    public mutating func replaceNote<T: NoteCollection>(at index: Int, with noteCollection: T) throws {
+        // validate they are the same duration
+        guard try note(at: index, sameDurationAs: [noteCollection]) else {
+            throw TupletError.replacementNotSameDuration
+        }
+        let flatIndex = flatIndexes[index]
+        let preservedTieStateCollection = try preserveTieStateForReplacement(in: index...index, with: noteCollection)
+        try replaceNote(at: flatIndex, with: preservedTieStateCollection)
     }
 
-    public mutating func replaceNote(at index: Int, with notes: [Note]) throws {
-        throw NSError()
+    public mutating func replaceNote<T: NoteCollection>(at index: Int, with noteCollections: [T]) throws {
+        // validate they are the same duration
```

Comment:
probably don't need this comment (same line 136). If it is confusing, then maybe change the method's name?